src/Neo/Ledger/MemoryPool.cs

```diff
@@ -374,28 +454,60 @@ internal void InvalidateVerifiedTransactions()
             _unsortedTransactions.Clear();
             VerificationContext = new TransactionVerificationContext();
             _sortedTransactions.Clear();
+            _conflicts.Clear();
         }
 
         // Note: this must only be called from a single thread (the Blockchain actor)
         internal void UpdatePoolForBlockPersisted(Block block, DataCache snapshot)
         {
+            var conflictingItems = new List<Transaction>();
             _txRwLock.EnterWriteLock();
             try
             {
+                HashSet<UInt256> conflicts = new HashSet<UInt256>();
                 // First remove the transactions verified in the block.
+                // No need to modify VerificationContext as it will be reset afterwards.
                 foreach (Transaction tx in block.Transactions)
                 {
-                    if (TryRemoveVerified(tx.Hash, out _)) continue;
-                    TryRemoveUnVerified(tx.Hash, out _);
+                    if (!TryRemoveVerified(tx.Hash, out _)) TryRemoveUnVerified(tx.Hash, out _);
+                    foreach (var h in tx.GetAttributes<Conflicts>().Select(a => a.Hash))
+                    {
+                        conflicts.Add(h);
+                    }
                 }
 
-                // Add all the previously verified transactions back to the unverified transactions
+                // Then remove the transactions conflicting with the accepted ones.
+                // No need to modify VerificationContext as it will be reset afterwards.
+                var persisted = block.Transactions.Select(t => t.Hash);
+                var stale = new List<UInt256>();
+                foreach (var item in _sortedTransactions)
+                {
+                    if (conflicts.Contains(item.Tx.Hash) || item.Tx.GetAttributes<Conflicts>().Select(a => a.Hash).Intersect(persisted).Count() > 0)
+                    {
+                        stale.Add(item.Tx.Hash);
+                        conflictingItems.Add(item.Tx);
+                    }
+                }
+                foreach (var h in stale)
+                {
+                    if (!TryRemoveVerified(h, out _)) TryRemoveUnVerified(h, out _);
+                }
```

Comment:
```suggestion
                var persisted = block.Transactions.Select(tx => tx.Hash).ToList();
                var stale = _sortedTransactions
                    .Where(item => conflicts.Contains(item.Tx.Hash)
                                   || item.Tx.GetAttributes<Conflicts>().Select(a => a.Hash).Intersect(persisted).Any())
                    .Select(item => item.Tx.Hash)
                    .ToList();
                conflictingItems.AddRange(stale.Select(h => _sortedTransactions.First(item => item.Tx.Hash == h).Tx));

                stale.Where(h => !TryRemoveVerified(h, out _)).ToList().ForEach(h => TryRemoveUnVerified(h, out _));
```

Reply:
Combine the two foreach loops into one using LINQ.
Avoid unnecessary list initialization.

Reply:
IIRC LINQ is somewhat slower, but I'm OK either way. @shargon?

Reply:
I prefer to avoid linq, it's slower

Reply:
Not a big deal, just ignore it.